Migrate to laminas/laminas-coding-standard v2

gennadigennadigennadi opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Bugfix        | yes
| QA            | yes

### Description

Started working on migration Code style to v2.

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

gennadigennadigennadi committed:
fix codestyle

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

tux-rampage commented on composer.json:
```diff
@@ -29,13 +29,13 @@
         "laminas/laminas-stdlib": "^3.2.1",
         "laminas/laminas-zendframework-bridge": "^0.4.5 || ^1.0",
-        "phpstan/phpstan-shim": "^0.10.5",
+        "phpstan/phpstan": "^0.12",
```
This is already targeted in PR #15 
Could you explain. why you changed it here? To me it seems that it is not needed to update the coding style.

tux-rampage replied:
Duplicate - I'll mark as resolved, sorry about this.

tux-rampage commented on phpstan.neon:
```diff
@@ -1,5 +1,5 @@
 parameters:
-    level: 7
+    level: 4
```
The same as for `composer.json`: This is already targeted in PR #15
Is it required for applying the CS?

tux-rampage replied:
Duplicate - I'll mark as resolved, sorry about this.

tux-rampage commented on src/CodeGenerator/InjectorGenerator.php:
```diff
@@ -39,70 +41,55 @@ class InjectorGenerator
+     * @param string|null                 $namespace Namespace to use for generated class; defaults
+     *                     to Laminas\Di\Generated.
+     * @param LoggerInterface|null        $logger An optional logger instance to log failures
+     *            and processed classes.
```
This looks a bit off @geerteltink and @michalbundyra - Is this how it should look like for multiline doc-tags?

tux-rampage replied:
Duplicate - I'll mark as resolved, sorry about this.

tux-rampage commented on composer.json:
```diff
@@ -29,13 +29,13 @@
         "laminas/laminas-stdlib": "^3.2.1",
         "laminas/laminas-zendframework-bridge": "^0.4.5 || ^1.0",
-        "phpstan/phpstan-shim": "^0.10.5",
+        "phpstan/phpstan": "^0.12",
```
This seems to be already addressed in PR #15 
Is this (Version bump for phpstan) needed for applying the CS or was it committed by accident? To me it seems like this is not necessary here.

tux-rampage commented on phpstan.neon:
```diff
@@ -1,5 +1,5 @@
 parameters:
-    level: 7
+    level: 4
```
Same as for `composer.json`. Is this required. It seems like a side effect of the phpstam version bump.
If not required, please revert

tux-rampage commented on src/CodeGenerator/InjectorGenerator.php:
```diff
@@ -39,70 +41,55 @@ class InjectorGenerator
+     * @param string|null                 $namespace Namespace to use for generated class; defaults
+     *                     to Laminas\Di\Generated.
+     * @param LoggerInterface|null        $logger An optional logger instance to log failures
+     *            and processed classes.
```
These comments look a bit off. Was this applied by the coding standard's auto-fixer?

gennadigennadigennadi replied:
yup. At least I would not Style code like this ;-)

tux-rampage commented on src/Config.php:
```diff
@@ -168,47 +166,45 @@ public function getTypePreference(string $type, ?string $context = null) : ?stri
 
         $preference = $this->types[$context]['preferences'][$type];
-        return ($preference != '') ? (string)$preference : null;
+        return $preference !== '' ? (string) $preference : null;
```
This introduces a BC break: When some Stringable (i.e. a class with a `__toString()` method) is placed in the config it will now change behavior. Prior this change it would be accepted and stringified. After this it becomes `null`.

gennadigennadigennadi replied:
This one was also made by php-cs. check if $preference is a class and has __toString then call __toString by ourselves?

gennadigennadigennadi replied:
@boesing and @tux-rampage this problem still has to be fixed!
What do you guys think how to solve it?

geerteltink replied:
@gennadigennadigennadi You can run `vendor/bin/phpcs -s` and check which rule is trying to change it. When you know, you can add `// phpcs:ignore some.phpcs.rule` in front of it so it won't get changed.

tux-rampage commented on src/Container/AutowireFactory.php:
```diff
@@ -41,18 +41,22 @@ private function getInjector(ContainerInterface $container)
     /**
      * Create an instance
+     *
+     * @return bool
```
This return type is incorrect, it shoud be `object`

tux-rampage commented on src/Container/AutowireFactory.php:
```diff
@@ -61,9 +65,11 @@ public function create(ContainerInterface $container, string $requestedName, ?ar
     /**
      * Make invokable and implement the laminas-service factory pattern
+     *
+     * @return bool
```
Same as above: Return type is `object`

tux-rampage commented on src/Resolver/AbstractInjection.php:
```diff
@@ -1,5 +1,5 @@
 <?php
-
+// phpcs:ignoreFile
```
Why should this file be ignored from CS?

gennadigennadigennadi replied:
Phpcs fails because of the deprecation warning.

boesing replied:
Maybe we should just disable the rule then instead of disabling the whole file from being checked.

tux-rampage commented on src/Resolver/DependencyResolverInterface.php:
```diff
@@ -32,24 +33,23 @@ public function setContainer(ContainerInterface $container);
-     *     preference for
-     * @param string $context The typename of the instance that is created or
+     * @param string      $type The type/class name of the dependency to resolve the
+     *          preference for
```
This looks also a bit off. Was this also applied by the auto-fixer?

gennadigennadigennadi replied:
Those multiline DocBlock got "fixed" automatically.

tux-rampage replied:
I see, thanks.

@boesing should it look like this. Maybe this is some edge case the auto-fixer ran into? Shouldn't this be just 4 spaces of indentation for each new line?

boesing replied:
> @boesing should it look like this. Maybe this is some edge case the auto-fixer ran into? Shouldn't this be just 4 spaces of indentation for each new line?

TBH - I dont know. I'd say, if the phpcbf fixes the code like this and it wasnt applied manually, we should keep this as is and either wait until there is a bugfix upstream. 🤷‍♂️ 
Feel free to report that.

tux-rampage replied:
Agree

tux-rampage commented on test/InjectorTest.php:
```diff
@@ -325,13 +328,13 @@ public function testComplexDeepDependencyConfiguration()
                 ],
-                'Level2.Alias' => [
-                    'typeOf' => TreeTestAsset\Level2::class,
+                'Level2.Alias'                        => [
```
This looks lie auto fixer. @geerteltink and @michalbundyra  This looks really wired and is hard to read for multi dimensional arrays. Is this the intended effect/style?

boesing replied:
As explained [here](https://github.com/laminas/laminas-di/pull/16#discussion_r499811386), we should keep this as is as it will probably re-formatted again by the beautifier.

tux-rampage commented on test/TestAsset/Constructor/OptionalArguments.php:
```diff
@@ -12,7 +12,7 @@
 class OptionalArguments
 {
-    public function __construct($foo = null, $bar = 'something')
+    public function __construct(?string $foo = null, string $bar = 'something')
```
This changes the constructor signature of this test asset. Consider using doc-block tags instead.

tux-rampage reviewed (changes requested):
Thank you very much for taking care of this. :+1:
There are some pieces that might need adjustments. Could you please check my feedback on the diff?

tux-rampage commented:
There are duplicate remarks, github was not showing them anymore after I temporarly navigated away from the review :facepalm: sorry about that.

The builds are failing for PHP `7.1` which is IMHO not a big deal - it's EOL anyways.

I'd suggest to wait for PR #15 (which is almost ready) to be merged since this PR must be rebased anyways afterward.

boesing reviewed (changes requested):
I think, the following TODOs are open:
- remove phpstan changes
- do not add native php typehints to `protected` or `public` methods (unless they're marked as `final`)  for classes which are not final as this will require a major version bump which should not be needed to just apply a new coding standard

Besides that, well done.

boesing commented on src/CodeGenerator/AbstractInjector.php:
```diff
@@ -19,32 +19,24 @@
      * {@inheritDoc}
      */
-    public function __construct(InjectorInterface $injector, ContainerInterface $container = null)
+    public function __construct(InjectorInterface $injector, ?ContainerInterface $container = null)
```
Please revert his change as it introduces BC break.

gennadigennadigennadi replied:
These additional ? were added by the phpcbf, but I added the others Typehints. As soon as #15 is merged I will revert some of the Typehints that are breaking changes.

tux-rampage replied:
Are you sure that this breaks BC? This is something the coding standard dictates. Afaik these types are already nullable   since they declare `null` as default value.

Edit: @boesing

boesing replied:
@gennadigennadigennadi right, in `__construct` thats not an issue (as it is treated different to other methods).

@tux-rampage nope, not sure. Actually, `__construct` signature changes do not have implications as tested [here](https://3v4l.org/bW2O8)

Methods other than `__construct` will lead to [this](https://3v4l.org/kmB34) error in classes which use inheritance

tux-rampage replied:
@boesing your example already leads to an error with the current codebase because it declares `null` as param default: https://3v4l.org/6GSPb

So `foo(type $var = null)` is an implicit `foo(?type = null)`. The coding style just enforces to make this obvious. So as far as I can tell this does not break BC or am I missing something else?

boesing replied:
I've re-checked this again and you are right... adding the nullable-flag to a typehint is not a BC break. I probably miss-understood something... Damn!

@gennadigennadigennadi sorry for bothering you with this.

boesing commented on src/CodeGenerator/InjectorGenerator.php:
```diff
@@ -39,70 +41,55 @@ class InjectorGenerator
         DependencyResolverInterface $resolver,
-        string $namespace = null,
-        LoggerInterface $logger = null
+        ?string $namespace = null,
```
Please revert his change as it introduces BC break.

gennadigennadigennadi replied:
those ? are added by cs-fix

boesing commented on src/Container/AutowireFactory.php:
```diff
@@ -41,18 +41,22 @@ private function getInjector(ContainerInterface $container)
+     * @return bool
      */
-    public function canCreate(ContainerInterface $container, $requestedName)
+    public function canCreate(ContainerInterface $container, string $requestedName)
```
Please revert his change as it introduces BC break.

boesing commented on src/Container/AutowireFactory.php:
```diff
@@ -61,9 +65,11 @@ public function create(ContainerInterface $container, string $requestedName, ?ar
+     * @return bool
      */
-    public function __invoke(ContainerInterface $container, $requestedName, array $options = null)
+    public function __invoke(ContainerInterface $container, string $requestedName, ?array $options = null)
```
Please revert his change as it introduces BC break.

gennadigennadigennadi replied:
I removed the string Typehint. The ? before array was added by cs-fix.

boesing commented on src/Container/ServiceManager/AutowireFactory.php:
```diff
@@ -21,18 +21,19 @@
     private $factory;
 
-    public function __construct(GenericAutowireFactory $factory = null)
+    public function __construct(?GenericAutowireFactory $factory = null)
```
Please revert his change as it introduces BC break.

gennadigennadigennadi replied:
those ? are added by cs-fix

boesing commented on src/Container/ServiceManager/AutowireFactory.php:
```diff
@@ -41,8 +42,11 @@ public function canCreate(ContainerInterface $container, $requestedName)
+     * @return bool
      */
-    public function __invoke(ContainerInterface $container, $requestedName, array $options = null)
+    public function __invoke(ContainerInterface $container, $requestedName, ?array $options = null)
```
Please revert his change as it introduces BC break.

boesing commented on src/Definition/RuntimeDefinition.php:
```diff
@@ -13,25 +13,25 @@
      * @param null|string[] $explicitClasses
      */
-    public function __construct(array $explicitClasses = null)
+    public function __construct(?array $explicitClasses = null)
```
Please revert his change as it introduces BC break.

boesing commented on src/Exception/ClassNotFoundException.php:
```diff
@@ -13,7 +13,7 @@
 class ClassNotFoundException extends DomainException implements ExceptionInterface
 {
-    public function __construct(string $classname, int $code = null, Throwable $previous = null)
+    public function __construct(string $classname, ?int $code = null, ?Throwable $previous = null)
```
Please revert his change as it introduces BC break.

boesing commented on src/Injector.php:
```diff
@@ -10,68 +10,68 @@
-        ContainerInterface $container = null,
-        Definition\DefinitionInterface $definition = null,
-        Resolver\DependencyResolverInterface $resolver = null
+        ?ConfigInterface $config = null,
```
Please revert his change as it introduces BC break.

boesing commented on src/Injector.php:
```diff
@@ -10,68 +10,68 @@
-        Definition\DefinitionInterface $definition = null,
-        Resolver\DependencyResolverInterface $resolver = null
+        ?ConfigInterface $config = null,
+        ?ContainerInterface $container = null,
```
Please revert his change as it introduces BC break.

boesing commented on src/Injector.php:
```diff
@@ -10,68 +10,68 @@
-        Resolver\DependencyResolverInterface $resolver = null
+        ?ConfigInterface $config = null,
+        ?ContainerInterface $container = null,
+        ?DefinitionInterface $definition = null,
```
Please revert his change as it introduces BC break.

gennadigennadigennadi replied:
those ? are added by cs-fix

boesing commented on src/Injector.php:
```diff
@@ -10,68 +10,68 @@
+        ?ConfigInterface $config = null,
+        ?ContainerInterface $container = null,
+        ?DefinitionInterface $definition = null,
+        ?DependencyResolverInterface $resolver = null
```
Please revert his change as it introduces BC break.

gennadigennadigennadi replied:
those ? are added by cs-fix

boesing commented on src/Resolver/DependencyResolver.php:
```diff
@@ -287,7 +285,7 @@ public function resolveParameters(string $requestedType, array $callTimeParamete
         foreach ($params as $paramInfo) {
             $name = $paramInfo->getName();
-            $type = $paramInfo->getType();
+            $type = $paramInfo->getType() === 'self' ? $requestedType : $paramInfo->getType();
```
We should only apply codestyle changes in this PR. 
If this would be an improvement, please create another PR with this change.

Please revert this change in this PR.

gennadigennadigennadi replied:
You are right. 
At first the phpcbf changes in the X.php/Y.php  the a Typehint to self and the ```$paramInfo->getType()``` returned the string `self` instead of the fqcn. Therefore I added the ```// phpcs:ignoreFile``` to X.php/Y.php I I forgot to revert these changes.

boesing commented on src/Resolver/ValueInjection.php:
```diff
@@ -120,12 +115,15 @@ public function toValue(ContainerInterface $container)
     {
         trigger_error(
-            __METHOD__ . ' is deprecated, please migrate to ' . __CLASS__ . '::toValue().',
+            __METHOD__ . ' is deprecated, please migrate to ' . self::class . '::toValue().',
```
Was that applied by the codestyle? Imho, we should stick with `__CLASS__` constant.

gennadigennadigennadi replied:
phpcbf did this.

tux-rampage commented:
@gennadigennadigennadi thanks for your patience. I've merged #15, could you please rebase your branch on `3.2.x` and address the feedback? :v:

gennadigennadigennadi committed:
reset previous changes

remove Typehints and add DocBlock
fix wrong DocBlock and remove Typehints
reset $type self logic

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

gennadigennadigennadi committed:
add one missing DocBlock

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

gennadigennadigennadi committed:
reset phpstan level

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

gennadigennadigennadi committed:
remove Typehints and add DocBlocks instead

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

gennadigennadigennadi commented:
I rebased yesterday and I am going to work the next few days on the feedback.

gennadigennadigennadi committed:
fix codestyle

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

gennadigennadigennadi committed:
add feedback

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

gennadigennadigennadi committed:
add feedback

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

gennadigennadigennadi committed:
update baseline

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

tux-rampage commented on src/Container/AutowireFactory.php:
```diff
@@ -41,29 +41,38 @@ private function getInjector(ContainerInterface $container)
+     * @return object
      */
-    public function create(ContainerInterface $container, string $requestedName, ?array $options = null)
+    public function create(ContainerInterface $container, $requestedName, ?array $options = null)
```
Even though this would not break BC, please don't remove the `string` typehint, that would remove type safety.

gennadigennadigennadi replied:
this one is a mistake, I thought this one was added by my commits

tux-rampage replied:
No worries, those slip through easily on such a large changeset. That's why we do reviews. :wink:

tux-rampage commented on src/LegacyConfig.php:
```diff
@@ -36,6 +49,10 @@ public function __construct($config)
 
+    /**
+     * @param mixed $parameters
+     * @return mixed
```
This should be an array of `mixed`

```suggestion
     * @return mixed[]
```

tux-rampage commented on src/LegacyConfig.php:
```diff
@@ -51,7 +68,7 @@ private function prepareParametersArray($parameters, string $class)
     }
 
-    private function configureInstance($config)
+    private function configureInstance(iterable $config)
```
This is a BC break. Please remove this type hint.

Edit: Event though this is private, we're not on psalm with full static analysis, yet and for phpstan there is a baseline in place. So it may break something undetected.

tux-rampage commented on test/Resolver/DependencyResolverTest.php:
```diff
@@ -132,7 +135,6 @@ private function mockDefinition(array $definition)
             $mock->method('getClassDefinition')
                 ->with($class)
                 ->willReturn($this->mockClassDefinition(
-                    $class,
```
Why do you remove this parameter?

gennadigennadigennadi replied:
https://github.com/laminas/laminas-di/blob/3.2.x/test/Resolver/DependencyResolverTest.php#L78
Ist not used inside mockClassDefinition. 
But as you said only code style changes, no refactorings. ;-). Im going to revert those last changes and I should be good.

tux-rampage reviewed (changes requested):
@gennadigennadigennadi Thank you very much for rebasing and adressing the feedback. There are still a view small things that would introduce breaking changes. Could you please check these?
Otherwise great work :+1:

gennadigennadigennadi committed:
revert previous bc

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

tux-rampage reviewed (approved):
LGTM. Great work, thank you very much. :+1: 

@boesing Can you please check if your feedback is was addressed or if there are any objections on your side?

boesing commented:
LGTM aswell. I definitely was wrong with that nullable flag. I was totally convinced it introduces BC break but it doesn't. Sorry for bothering you two...
